dashboard/dashboard.go

```diff
@@ -0,0 +1,281 @@
+		if err = template.Must(template.New("").Parse(string(tmpl))).Execute(website, map[string]interface{}{
+			"procSampleLimit": procSampleLimit,
+			"memSampleLimit":  memSampleLimit,
+		}); err != nil {
+			log.Crit("Failed to render the dashboard template", "err", err)
+		}
+
+		dashboard.index = website.Bytes()
+		return dashboard, nil
+	}
+
+	//TODO case: DashboardAssetsFlag is set
+	//dashboard.index = ioutil.ReadFile()
+	return dashboard, nil
+}
+
+func (db *dashboard) Protocols() []p2p.Protocol { return nil }
+
+func (db *dashboard) APIs() []rpc.API { return nil }
+
+func (db *dashboard) Start(server *p2p.Server) error {
+	log.Trace("Start() called", "config", db.config)
+
+	go db.collectData()
+
+	http.HandleFunc("/", db.webHandler)
+	http.Handle("/api", websocket.Handler(db.apiHandler))
+
+	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", db.config.Host, db.config.Port))
+	if err != nil {
+		return err
+	}
+	db.listener = listener
+
+	go func() {
+		log.Trace("Starting server...")
+
+		if err := http.Serve(listener, nil); err != nil {
+			log.Warn("Server failed", "err", err)
+		}
+	}()
+
+	return nil
+}
+
+func (db *dashboard) Stop() error {
+	log.Trace("Terminating dashboard...")
+
+	var err error
+
+	db.lock.Lock()
+	if err = db.listener.Close(); err != nil {
+		log.Warn("Failed to close listener", "err", err)
+	}
+
+	for _, c := range db.conns {
+		if err := c.conn.Close(); err != nil {
+			c.logger.Warn("Failed to close connection", "err", err)
+		}
+	}
+	db.conns = db.conns[:0]
+	db.lock.Unlock()
+
+	return err
+}
+
+// webHandler handles all non-api requests, simply flattening and returning the dashboard website.
+func (db *dashboard) webHandler(w http.ResponseWriter, r *http.Request) {
+	log.Trace("webHandler() called")
+
+	//TODO not only index
+	w.Write(db.index)
+}
+
+// apiHandler handles requests for dashboard
+func (db *dashboard) apiHandler(conn *websocket.Conn) {
+	log.Trace("apiHandler() called")
+
+	client := &client{
+		conn:   conn,
+		logger: log.New("id", atomic.AddUint32(&nextId, 1)),
+	}
+
+	// Start tracking the connection and drop at connection loss
+	db.lock.Lock()
+	db.conns = append(db.conns, client)
+	db.lock.Unlock()
+
+	defer func() {
+		client.logger.Trace("Connection interrupted")
+
+		db.lock.Lock()
+		for i, c := range db.conns {
+			if c.conn == client.conn {
+				if err := c.conn.Close(); err != nil {
+					c.logger.Warn("Failed to close connection", "err", err)
+				}
+				db.conns = append(db.conns[:i], db.conns[i+1:]...)
+				break
+			}
+		}
+		db.lock.Unlock()
+	}()
+
+	db.sendHistory(client)
+
+	for {
+		var msg struct {
+			text string `json:"text"`
+		}
+		if err := websocket.JSON.Receive(conn, &msg); err != nil {
+			client.logger.Warn("Receive failed", "err", err)
+			return
+		}
+		// Ignore any message
+	}
+}
+
+// collectData collects the required data to plot on the dashboard
+func (db *dashboard) collectData() {
+	log.Trace("collectData() called")
+
+	for {
```

Comment:
This is an infinite loop that never terminates (i.e. a goroutine leak). You need to ensure that calling dashboard.Stop terminates any running goroutines before the system assumes the dashboard terminated.

The simplest way to achieve this is to add a quit channel `chan chan error` to the dashboard, and when you terminate the dashboard, you send a `chan error` into it, and read back an `error` from that. In this goroutine then instead of `time.Sleep` at the bottom, you use:

```
select {
  case errc := <-db.quit:
    errc <- nil
    return
  case <-time.After(db.config.Refresh):
}
```